hpx/hpx_init.hpp

```diff
@@ -14,55 +14,222 @@
 
 #include <hpx/config.hpp>
 #include <hpx/hpx_finalize.hpp>
+#include <hpx/hpx_init_params.hpp>
 #include <hpx/hpx_suspend.hpp>
 #include <hpx/program_options.hpp>
+#include <hpx/resource_partitioner/partitioner_fwd.hpp>
 #include <hpx/runtime/runtime_mode.hpp>
 #include <hpx/runtime/shutdown_function.hpp>
 #include <hpx/runtime/startup_function.hpp>
 #include <hpx/functional/function.hpp>
 
 #include <cstddef>
+#include <memory>
 #include <string>
 #include <vector>
 
-#ifndef DOXYGEN
-#if !defined(HPX_MAIN_EXPORT)
-#define HPX_MAIN_EXPORT /**/
-#endif
-
-///////////////////////////////////////////////////////////////////////////////
-// One of these functions must be implemented by the application for the
-// console locality.
-int hpx_main();
-HPX_MAIN_EXPORT int hpx_main(int argc, char** argv);
-int hpx_main(hpx::program_options::variables_map& vm);
-#endif
-
-///////////////////////////////////////////////////////////////////////////////
-/// \cond NOINTERNAL
-namespace hpx_startup
-{
-    // As an alternative, the user can provide a function hpx_startup::user_main,
-    // which is semantically equivalent to the plain old C-main.
-    int user_main();
-    int user_main(int argc, char** argv);
-}
-/// \endcond
 
 ///////////////////////////////////////////////////////////////////////////////
 /// \namespace hpx
 namespace hpx
 {
-    namespace detail
-    {
-        HPX_EXPORT void on_exit() noexcept;
-        HPX_EXPORT void on_abort(int signal) noexcept;
-    }
+    /// \brief Main entry point for launching the HPX runtime system.
+    ///
+    /// This is the main entry point for any HPX application. This function
+    /// (or one of its overloads below) should be called from the users `main()`
+    /// function. It will set up the HPX runtime environment and schedule the
+    /// function given by \p f as a HPX thread. This overload will not call
+    /// `hpx_main`.
+    ///
+    ///
+    /// \param f            [in] The function to be scheduled as an HPX
+    ///                     thread. Usually this function represents the main
+    ///                     entry point of any HPX application. If \p f is
+    ///                     `nullptr` the HPX runtime environment will be started
+    ///                     without invoking \p f.
+    /// \param argc         [in] The number of command line arguments passed
+    ///                     in \p argv. This is usually the unchanged value as
+    ///                     passed by the operating system (to `main()`).
+    /// \param argv         [in] The command line arguments for this
+    ///                     application, usually that is the value as passed
+    ///                     by the operating system (to `main()`).
+    /// \param params       [in] The parameters to the init function of type
+    ///                     `hpx::init_params`. Usually the
```

Comment:
I think we can keep this a bit shorter and just refer to `init_params` documentation.

Reply:
Yep !